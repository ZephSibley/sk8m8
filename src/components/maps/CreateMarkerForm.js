import React, { useState, useEffect } from 'react';
import TextField from '@material-ui/core/TextField';
import InputLabel from '@material-ui/core/InputLabel';
import Select from '@material-ui/core/Select';
import Button from '@material-ui/core/Button';
import FormControl from '@material-ui/core/FormControl';

import fetchLocationTypes from '../../utils/map/fetchLocationTypes';
import formStyles from '../../styles/forms';

const CreateMarkerForm = props => {
    // prop: location; array, latlng 

    const [textFieldValues, setTextFieldValues] = useState({
        name: '',
        description: '',
    });
    const [locationTypes, updateLocationTypes] = useState(
        <option>Loading...</option>
    );

    const handleChange = name => event => {
        setTextFieldValues({ ...textFieldValues, [name]: event.target.value });
    };

    useEffect(() => {
        fetchLocationTypes()
        .then(data => {
            const options = data.map((type) => 
                    <option key={type}>{type}</option>
            );
            options.unshift(<option key="blank" ></option>)
            updateLocationTypes(options);
        }).catch(
            updateLocationTypes(
                <option>Something went wrong D:</option>
            )
        );
    }, []);

    const formClasses = formStyles()
    return (
        <div>
            <h4 className={formClasses.embeddedFormHeader}>
                Make your mark
            </h4>
            <form className={formClasses.form} action="/" method="post">
                <input type='hidden' value={props.location} id='marker-location' name='marker-location' />
                <TextField
                    required
                    id='marker-name'
                    label="Name"
                    value={textFieldValues.name}
                    onChange={handleChange('name')}
                    margin="normal"
                />
                <FormControl>
                    <InputLabel htmlFor="location-type">Location Type</InputLabel>
                    <Select
                        required
                        native
                        inputProps={{ name: 'location-type', id: 'location-type' }}
                        style={{width: 200}}
                    >
                        {locationTypes}
                    </Select>
                </FormControl>
                <TextField
                    id='marker-creator'
                    label='Creator'
                    value={'username'/*get username*/}
                    margin='normal'
                    InputProps={{
                        readOnly: true,
                    }}
                />
                <input
                    required
                    style={{ display: 'none' }}
                    accept="video/*"
                    id="marker-video"
                    type="file"
                />
                <label htmlFor="marker-video">
                    <Button
                        variant="outlined"
                        component="span"
                        color="secondary"
                        className={formClasses.submit}
                    >
                        Choose Video
                    </Button>
                </label>
                <input
                    type="hidden"
                    id="marker-location"
                    name="marker-location"
                    value={props.location}
                />
                <input
                    style={{ display: 'none' }}
                    id="marker-submit"
                    multiple
                    type="submit"
                />
                <label htmlFor="marker-submit">
                    <Button
                        variant="contained"
                        color="primary"
                        component="span"
                        className={formClasses.submit}
                    >
                        Create
                    </Button>
                </label>
            </form>
        </div>
    )
}

export default CreateMarkerForm;