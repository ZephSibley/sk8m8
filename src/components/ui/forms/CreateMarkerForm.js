import React, { useState, useEffect } from 'react';
import { Formik, Field } from 'formik';
import *  as Yup from 'yup';
import TextField from '@material-ui/core/TextField';
import InputLabel from '@material-ui/core/InputLabel';
import Select from '@material-ui/core/Select';
import FormHelperText from '@material-ui/core/FormHelperText';
import Button from '@material-ui/core/Button';
import FormControl from '@material-ui/core/FormControl';

import VideoInput from '../inputs/VideoInput';
import formStyles from '../../../styles/forms';
import BackendValidationError from '../text/BackendValidationError';
import { Typography } from '@material-ui/core';


const CreateMarkerForm = props => {
    // prop: location; array, latlng
    // prop: requests; http client

    const formClasses = formStyles()

    const [locationTypes, updateLocationTypes] = useState(
        <option>Loading...</option>
    );
    const [success, setSuccess] = useState(false);
    const [submitError, setSubmitError] = useState([]);

    useEffect(() => {
        props.requests.get(
            `${process.env.REACT_APP_ENDPOINT}mapmarker/locationTypes`,
            {
                headers: {
                    'Authorization': `Bearer ${sessionStorage.getItem('token')}`
                }
            }
        ).then(response => {
            const options = response.data.map((type, index) =>
                <option key={index}>{type}</option>
            );
            options.unshift(<option key="blank" ></option>)
            updateLocationTypes(options);
        }).catch( updateLocationTypes(
                <option>Something went wrong D:</option>
            ));
    }, [props.requests]);

    const validationSchema = Yup.object().shape({
        name: Yup.string()
            .required("Please give your marker a name")
            .min(3, "min. 3 characters"),
        category: Yup.string()
            .required("What kind of location are you marking?"),
        /* VIDEO UPLOAD FEATURE
        video: Yup.mixed()
            .required("We need a video!")
            .test(
                "fileSize",
                "Your video is too big :(",
                value => value && value.size <= 262144000
            )
        */
    })

    return (
        <div>
            <h4 className={formClasses.embeddedFormHeader}>
                Make your mark
        </h4>
            <Formik
                initialValues={{
                    name: '',
                    category: '',
                    video: null,
                    latitude: props.location[0],
                    longitude: props.location[1],
                }}
                validationSchema={validationSchema}
                onSubmit={(values, { setSubmitting }) => {
                    setSubmitting(true);
                    console.log(values)
                    props.requests.post(
                        `${process.env.REACT_APP_ENDPOINT}mapmarker/create`,
                        values,
                        {
                            headers: {
                                'Authorization': `Bearer ${sessionStorage.getItem('token')}`
                            }
                        }
                    ).then(response =>
                        setSuccess(true)
                    ).catch(err => {
                        setSubmitError(err.message || Object.values(JSON.parse(err)));
                        setSubmitting(false);
                    });
                }}

                render={({
                    values,
                    errors,
                    touched,
                    handleChange,
                    setFieldValue,
                    handleBlur,
                    isValid,
                    handleSubmit,
                    isSubmitting,
                }) => (
                        <form
                            onSubmit={handleSubmit}
                            className={formClasses.form}
                        >
                            <FormHelperText>
                                Mark this location on the map for others to see!<br />
                                {/*Include a video but please keep it under 10 seconds. VIDEO UPLOAD FEATURE */}
                            </FormHelperText>
                            <TextField
                                name='name'
                                label='name'
                                onChange={handleChange}
                                onBlur={handleBlur}
                                value={values.name}
                                margin='normal'
                                error={errors.name && touched.name}
                                helperText={
                                    errors.name && touched.name ?
                                        errors.name : ''
                                }
                            />
                            <FormControl margin='normal' style={{ maxWidth: 200 }}>
                                <InputLabel htmlFor="category">Location Type</InputLabel>
                                <Select
                                    native
                                    inputProps={{ name: 'category', id: 'category' }}
                                    style={{ width: 200, }}
                                    value={values.category}
                                    onChange={handleChange}
                                    onBlur={handleBlur}
                                >
                                    {locationTypes}
                                </Select>
                                {errors.category &&
                                    touched.category &&
                                    <FormHelperText id="category_helper_text" error={true}>
                                        {errors.category}
                                    </FormHelperText>
                                }
                            </FormControl>
                            {process.env.REACT_APP_video_FEATURE ?
                            <Field
                                name="video"
                                component={VideoInput}
                                title="Upload"
                                setFieldValue={setFieldValue}
                                errorMessage={errors.video ?
                                    errors.video : ''
                                }
                                touched={touched.video}
                                onBlur={handleBlur}
                            /> :
                            <Typography color="textSecondary">Video upload coming soon!</Typography>
                            }
                            <BackendValidationError resp={submitError} />
                            <Button
                                variant="contained"
                                color="primary"
                                component="button"
                                type="submit"
                                margin='normal'
                                className={formClasses.submit}
                                disabled={isSubmitting || !isValid}
                            >
                                Create
                    </Button>
                        </form>
                    )}
            />
        </div>
    )
}

export default CreateMarkerForm;