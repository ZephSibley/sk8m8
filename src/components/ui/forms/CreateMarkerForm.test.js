import React from 'react';
import renderer from 'react-test-renderer';
import CreateMarkerForm from './CreateMarkerForm';

test('Basic Rendering', () => {
    const component = renderer.create(
        <CreateMarkerForm />
    )
    let tree = component.toJSON
    expect(tree).toMatchSnapshot();
})