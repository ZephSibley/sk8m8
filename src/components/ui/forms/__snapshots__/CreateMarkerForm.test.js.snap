// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic Rendering 1`] = `
<div>
  <h4
    className="makeStyles-embeddedFormHeader-9"
  >
    Make your mark
  </h4>
  <form
    className="makeStyles-form-1"
    onSubmit={[Function]}
  >
    <p
      className="MuiFormHelperText-root"
    >
      Mark YOUR CURRENT LOCATION on the map for others to see!
      <br />
    </p>
    <div
      className="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
        data-shrink={false}
      >
        name
      </label>
      <div
        className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          className="MuiInputBase-input MuiInput-input"
          disabled={false}
          name="name"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          type="text"
          value=""
        />
      </div>
      
    </div>
    <div
      className="MuiFormControl-root MuiFormControl-marginNormal"
      style={
        Object {
          "maxWidth": 200,
        }
      }
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
        data-shrink={false}
        htmlFor="category"
      >
        Location Type
      </label>
      <div
        className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
        onClick={[Function]}
        style={
          Object {
            "width": 200,
          }
        }
      >
        <select
          aria-invalid={false}
          className="MuiSelect-root MuiSelect-select MuiInputBase-input MuiInput-input MuiInputBase-inputSelect"
          disabled={false}
          id="category"
          multiple={false}
          name="category"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          value=""
        >
          <option>
            Loading...
          </option>
        </select>
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root MuiSelect-icon"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
    </div>
    <p
      className="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
    >
      Video upload coming soon!
    </p>
    <pre
      className="MuiFormHelperText-root Mui-error"
    />
    <button
      className="MuiButtonBase-root MuiButton-root makeStyles-submit-5 MuiButton-contained MuiButton-containedPrimary Mui-disabled Mui-disabled"
      disabled={true}
      margin="normal"
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={-1}
      type="submit"
    >
      <span
        className="MuiButton-label"
      >
        Create
      </span>
    </button>
  </form>
</div>
`;
